gnucash/gnome-utils/gnc-plugin.c

```diff
@@ -231,155 +223,266 @@ gnc_plugin_get_name (GncPlugin *plugin)
 
-    g_return_val_if_fail (ui_merge, 0);
-    g_return_val_if_fail (action_group, 0);
-    g_return_val_if_fail (filename, 0);
+    g_return_if_fail (display_item_hash != NULL);
+    g_return_if_fail (menubar != NULL);
 
-    ENTER("ui_merge %p, action_group %p, filename %s",
-          ui_merge, action_group, filename);
-    gtk_ui_manager_insert_action_group (ui_merge, action_group, 0);
+    ENTER("display_item_hash %p, menubar %p", display_item_hash, menubar);
 
-    pathname = gnc_filepath_locate_ui_file (filename);
-    if (pathname == NULL)
+    display_list = gnc_menu_get_items (menubar);
+
+    for (GList *ptr = display_list; ptr; ptr = g_list_next (ptr))
     {
-        LEAVE("fail");
-        return 0;
-    }
+        GncDisplayItem *gdi;
+        GtkWidget *item = GTK_WIDGET(ptr->data);
+        const gchar *action_name = (const gchar*)g_object_get_data (G_OBJECT(item), "myaction-name");
 
-    merge_id = gtk_ui_manager_add_ui_from_file (ui_merge, pathname, &error);
-    DEBUG("merge_id is %d", merge_id);
+        PINFO("action_name is '%s'", action_name);
 
-    g_assert(merge_id || error);
-    if (merge_id)
-    {
-        gtk_ui_manager_ensure_update (ui_merge);
+        if (!action_name)
+            continue;
+
+        gdi = (GncDisplayItem*)g_hash_table_lookup (display_item_hash, action_name);
+
+        if (!gdi)
+            continue;
+
+        PINFO("action_name is '%s', label '%s'", action_name, _(gdi->label));
+
+        gtk_menu_item_set_label (GTK_MENU_ITEM(item), _(gdi->label));
+        gtk_menu_item_set_use_underline (GTK_MENU_ITEM(item), TRUE);
+
+        if (gdi->accelerator)
+        {
+            GtkWidget *child = gtk_bin_get_child (GTK_BIN (item));
+            guint  accel_key = 0;
+            GdkModifierType accel_mods;
+
+            gtk_accelerator_parse (gdi->accelerator, &accel_key, &accel_mods);
+
+            if (accel_key > 0)
+                gtk_accel_label_set_accel (GTK_ACCEL_LABEL(child), accel_key, accel_mods);
+        }
+        if (gdi->tooltip)
+        {
+            gtk_widget_set_tooltip_text (item, gdi->tooltip);
+            gnc_menu_item_setup_tooltip_to_statusbar_callback (item, statusbar);
+        }
     }
-    else
+    g_list_free (display_list);
+    LEAVE("");
+}
+
+
+void
+gnc_plugin_update_display_toolbar_items (GHashTable *display_item_hash,
+                                         GtkWidget *toolbar,
+                                         GtkWidget *statusbar)
+{
+    g_return_if_fail (display_item_hash != NULL);
+    g_return_if_fail (toolbar != NULL);
+
+    ENTER("display_item_hash %p, toolbar %p", display_item_hash, toolbar);
+
+    for (gint i = 0; i < gtk_toolbar_get_n_items (GTK_TOOLBAR(toolbar)); i++)
     {
-        g_critical("Failed to load ui file.\n  Filename %s\n  Error %s",
-                   filename, error->message);
-        g_error_free(error);
-    }
+        GtkToolItem *item = gtk_toolbar_get_nth_item (GTK_TOOLBAR(toolbar), i);
```

Comment:
As you iterate over all the child items in the toolbar it may be more efficient to use gtk_container_foreach rather than gtk_toolbar_get_nth_item.

Reply:
OK, will have a look.

Reply:
I am going to leave this as is for now, I think I did it this way as I would need to create a structure to pass the parameters needed in the container for each and as there were only small number of items did no think it worth doing but may come back to it once other changes are possibly made.